[docs]: Update FFI chapter (#209)
hyperledger/iroha-2-docs: src/guide/advanced/ffi.md

## FFI Binding Generation

The `iroha_ffi` crate is used to generate functions that are callable via
FFI. Given `Rust` structs and methods, they generate the `unsafe` code that
you would need in order to cross the linking boundary.

> A Rust type is converted into a robust `repr(C)` type that can cross FFI

Reviewer: ```suggestion
A Rust type is converted into a robust `repr(C)` type that can cross the FFI
```

boundary with `FfiType::into_ffi`. This goes the other way around as well:
FFI `ReprC` type is converted into a `Rust` type via
`FfiType::try_from_ffi`.

::: warning

Note that the opposite conversion is fallible and can cause undefined
behaviour. While we can make the best effort to avoid the most obvious

> mistakes, you must ensure the correctness of the program on your end.

Reviewer: ```suggestion
mistakes, you must ensure the program's correctness on your end.
```

| ------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `ReprC`      | This trait represents a robust type that conforms to C ABI. The type can be safely shared across FFI boundaries.                                                                |
| `FfiType`    | This trait defines a corresponding `ReprC` type for a given `Rust` type. The defined `ReprC` type is used in place of the `Rust` type in the API of the generated FFI function. |
| `FfiConvert` | This trait defines two methods `into_ffi` and `try_from_ffi` that are used to perform the conversion of the `Rust` type to or from `ReprC` type.                                |

Note that there is no ownership transfer over FFI except for opaque pointer
types. All other types that carry ownership, such as `Vec<T>`, are cloned.

### Name Mangling

Note the use of double underscores in generated names of FFI objects:

- For the `inherent_fn` method defined on the `StructName` struct, the FFI
  name would be `StructName__inherent_fn`.

> - For the `MethodName` method from the `TraitName` trait in `StructName`

Reviewer: ```suggestion
- For the `MethodName` method from the `TraitName` trait in the `StructName`
```

  struct, the FFI name would be `StructName__TraitName__MethodName`.

> - For setting the `field_name` field in `StructName` struct, the FFI

Reviewer: ```suggestion
- To set the `field_name` field in the `StructName` struct, the FFI
```

  function name would be `StructName__set_field_name`.

> - For getting the `field_name` field in `StructName` struct, the FFI

Reviewer: ```suggestion
- To get the `field_name` field in the `StructName` struct, the FFI
```

  function name would be `StructName__field_name`.

> - For getting the mutable `field_name` field in `StructName` struct, the

Reviewer: ```suggestion
- To get the mutable `field_name` field in the `StructName` struct, the
```
